azkaban-common/src/main/java/azkaban/executor/container/ContainerizedDispatchImpl.java

```diff
@@ -0,0 +1,415 @@
+  }
+
+  /**
+   * This method will shutdown the queue processor thread. It won't pick up executions from queue
+   * for dispatch after this. This method will be called when webserver will shutdown.
+   */
+  @Override
+  public void shutdown() {
+    logger.info("Shutting down queue processor thread for containerized dispatch implementation.");
+    if (null != this.queueProcessor) {
+      this.queueProcessor.shutdown();
+    }
+  }
+
+  /**
+   * This method is used to enable queue processor thread. It will resume dispatching executions.
+   */
+  @Override
+  public void enableQueueProcessorThread() {
+    this.queueProcessor.setActive(true);
+  }
+
+  /**
+   * This method is used to disable queue processor thread. It will stop dispatching executions.
+   */
+  @Override
+  public void disableQueueProcessorThread() {
+    this.queueProcessor.setActive(false);
+  }
+
+  public State getQueueProcessorThreadState() {
+    return this.queueProcessor.getState();
+  }
+
+  public boolean isQueueProcessorThreadActive() {
+    return this.queueProcessor.isActive();
+  }
+
+  /**
+   * This QueueProcessorThread will fetch executions from database in batch/single execution at a
+   * time and dispatch it in container.
+   */
+  private class QueueProcessorThread extends Thread {
+
+    private final Props azkProps;
+    private volatile boolean shutdown = false;
+    private volatile boolean isActive = true;
+    private ExecutorService executorService;
+    private ExecutorLoader executorLoader;
+    private boolean executionsBatchProcessingEnabled;
+    private int executionsBatchSize;
+    private static final long QUEUE_PROCESSOR_WAIT_IN_MS = 1000;
+
+    public QueueProcessorThread(final Props azkProps, final ExecutorLoader executorLoader) {
+      this.azkProps = azkProps;
+      this.executorLoader = executorLoader;
+      setActive(
+          this.azkProps.getBoolean(Constants.ConfigurationKeys.QUEUEPROCESSING_ENABLED, true));
+      this.executionsBatchProcessingEnabled = azkProps
+          .getBoolean(ContainerizedExecutionManagerProperties.CONTAINERIZED_EXECUTION_BATCH_ENABLED,
+              false);
+      this.executionsBatchSize =
+          azkProps
+              .getInt(ContainerizedExecutionManagerProperties.CONTAINERIZED_EXECUTION_BATCH_SIZE,
+                  10);
+
+      this.executorService = Executors.newFixedThreadPool(azkProps.getInt(
+          ContainerizedExecutionManagerProperties.CONTAINERIZED_EXECUTION_PROCESSING_THREAD_POOL_SIZE,
+          10));
+      this.setName("Containerized-QueueProcessor-Thread");
+    }
+
+    @Override
+    public void run() {
+      // Loops till QueueProcessorThread is shutdown
+      while (!this.shutdown) {
+        synchronized (this) {
+          try {
+            // Start processing queue if active, otherwise wait for sometime
+            if (this.isActive) {
+              processQueuedFlows();
+            }
+            wait(QUEUE_PROCESSOR_WAIT_IN_MS);
```

Comment:
This wait is unnecessary. `RateLimiter.acquire()` will block as needed.

Reply:
Yes. I agree. Some test cases are failing when I remove wait. I am investigating it.

Reply:
But rate limiter is anyway needed as if we fetch executions in batch then rate limiter will take care of the rate at which containers will be created. For now, I am keeping this as it is helping to fix unit tests. I will add TODO to remove it and take care of it later on.

Reply:
I have fixed test cases and removed synchronize block and wait.